modules/mtl/arrow-mtl-data/src/main/kotlin/arrow/mtl/AccumT.kt

```diff
@@ -0,0 +1,211 @@
+package arrow.mtl
+
+import arrow.Kind
+import arrow.core.Either
+import arrow.core.Tuple2
+import arrow.core.identity
+import arrow.core.toT
+import arrow.higherkind
+import arrow.typeclasses.Applicative
+import arrow.typeclasses.Functor
+import arrow.typeclasses.Monad
+import arrow.typeclasses.Monoid
+
+/**
+ * Alias that represent stateful computation of the form `(S) -> Tuple2<S, A>` with a result in certain context `F`.
+ */
+typealias AccumTFun<S, F, A> = (S) -> Kind<F, Tuple2<S, A>>
+
+/**
+ * Alias that represents wrapped stateful computation in context `F`.
+ */
+typealias AccumTFunOf<S, F, A> = Kind<F, AccumTFun<S, F, A>>
+
+/**
+ * `AccumT<S, F, A>` is a stateful computation within a context `F` yielding
+ * a value of type `A`. AccumT provides append-only accumulation during the computation.
+ * For more general access, use `StateT` instead.
+ *
+ * When flatMapping the resulting state of subcomputations is combined using `Monoid<S>`
+ *
+ * @param F the context that wraps the stateful computation.
+ * @param S the state we are performing computation upon.
+ * @param A current value of computation.
+ * @param accumF the stateful computation that is wrapped and managed by `AccumT`
+ */
+@higherkind
+data class AccumT<S, F, A>(val accumT: AccumTFunOf<S, F, A>) : AccumTOf<S, F, A> {
+
+  companion object {
+
+    fun <S, F, A> just(MS: Monoid<S>, MF: Monad<F>, a: A): AccumT<S, F, A> =
+      AccumT(MF.just { w: S -> MF.just(MS.empty() toT a) })
+
+    operator fun <S, F, A> invoke(AF: Applicative<F>, accumTFun: AccumTFun<S, F, A>): AccumT<S, F, A> =
+      AF.run {
+        AccumT(just(accumTFun))
+      }
+
+    fun <S, F, A, B> tailRecM(
+      MF: Monad<F>,
+      a: A,
+      fa: (A) -> Kind<AccumTPartialOf<S, F>, Either<A, B>>
+    ): AccumT<S, F, B> = MF.run {
+      AccumT(MF) { s: S ->
+        tailRecM(Tuple2(s, a)) { (s1, a0) ->
+          fa(a0).fix().runAccumT(MF, s1).map { (s2, ab) ->
+            ab.bimap({ a1 -> s2 toT a1 }, { b -> s2 toT b })
+          }
+        }
+      }
+    }
+
+    fun <F, S, A> liftF(AF: Applicative<F>, fa: Kind<F, A>): AccumT<S, F, A> = AF.run {
+      AccumT(AF) { s -> fa.map { a -> Tuple2(s, a) } }
```

Comment:
This differs from the MonadTrans definition and causes the tests to fail. The tuple should contain `mempty` and not `s`.